### PR TITLE
AL/MIN-183. MNT integration.

### PR DESCRIPTION
**Pull request by lut1k**
**Description:**

1. Added extrinsic `claim_mnt()`. Added benchamrking test for this extrinsic.
2. Added new trait: `pub trait MntManager<AccountId>{}`.
3. MNT tokens distribution is integrated into the protocol.
4. Fixed benchmarking tests for minterest-protocol pallet.

**Changes in storage or API:**

The storage has not changed.
API: Added new extrinsic `claim_mnt()`;

**Checklist**

- [x] Code is covered with tests
- [x] Code is covered with benchmarks
- [x] Fresh benchmark results included

### CODE REVIEW

File: pallets/mnt-token/src/lib.rs [hunk 1]
```diff
@@ -285,67 +282,146 @@ pub mod module {
...
 
-		let delta_index = pool_borrow_state
-			.mnt_distribution_index
-			.checked_sub(&borrower_index)
+	/// Transfer MNT tokens to user balance if they are above the threshold.
+	/// Otherwise, put them into internal storage.
+	///
+	/// - `user`: MNT tokens recipient.
+	/// - `user_accrued`: The total amount of accrued tokens.
+	/// - `distribute_all`:
```
Comment (dev-vadim) [inline pallets/mnt-token/src/lib.rs:346]: Please add comment

File: pallets/controller/src/lib.rs [hunk 2]
```diff
@@ -562,6 +565,9 @@ impl<T: Config> Pallet<T> {
 					.map_err(|_| Error::<T>::HypotheticalLiquidityCalculationError)?;
 
 			ensure!(shortfall.is_zero(), Error::<T>::InsufficientLiquidity);
+
+			T::MntManager::update_mnt_supply_index(underlying_asset)?;
+			T::MntManager::distribute_supplier_mnt(underlying_asset, redeemer, false)?;
```
[the remark this answers is not in the document]
Comment (lut1k) [reply pallets/controller/src/lib.rs:570]: Fixed

File: pallets/controller/src/lib.rs [hunk 3]
```diff
@@ -582,6 +588,9 @@ impl<T: Config> Pallet<T> {
 
 		ensure!(shortfall.is_zero(), Error::<T>::InsufficientLiquidity);
 
+		T::MntManager::update_mnt_borrow_index(underlying_asset)?;
+		T::MntManager::distribute_borrower_mnt(underlying_asset, who, false)?;
```
[the remark this answers is not in the document]
Comment (lut1k) [reply pallets/controller/src/lib.rs:592]: Fixed

File: pallets/mnt-token/src/lib.rs (hunk 1 above)
Comment (lut1k) [reply pallets/mnt-token/src/lib.rs:346]: Fixed

File: pallets/minterest-protocol/src/lib.rs [hunk 4]
```diff
@@ -466,6 +483,9 @@ impl<T: Config> Pallet<T> {
 
 		<Controller<T>>::accrue_interest_rate(underlying_asset).map_err(|_| Error::<T>::AccrueInterestFailed)?;
 
+		<T as module::Config>::MntManager::update_mnt_supply_index(underlying_asset)?;
+		<T as module::Config>::MntManager::distribute_supplier_mnt(underlying_asset, who, false)?;
```
Comment (dev-vadim) [inline pallets/minterest-protocol/src/lib.rs:487]:
```suggestion
		T::MntManager::update_mnt_supply_index(underlying_asset)?;
		T::MntManager::distribute_supplier_mnt(underlying_asset, who, false)?;
```
And the same for other places

File: runtime/src/benchmarking/minterest_protocol.rs [hunk 5]
```diff
@@ -1,54 +1,62 @@
...
 	// set balance for users
-	set_balance(MDOT, &borrower, 10_000 * DOLLARS)?;
-	set_balance(METH, &borrower, 10_000 * DOLLARS)?;
-	set_balance(MKSM, &borrower, 10_000 * DOLLARS)?;
-	set_balance(MBTC, &borrower, 30_000 * DOLLARS)?;
-	set_balance(MDOT, &lender, 20_000 * DOLLARS)?;
+	EnabledWrappedTokensId::get()
+		.into_iter()
+		.try_for_each(|token_id| -> Result<(), &'static str> {
+			set_balance(token_id, borrower, 10_000 * DOLLARS)?;
```
Comment (Apollo50) [inline runtime/src/benchmarking/minterest_protocol.rs:35]: With this setup we have got exchange rate in BTC pool different from 1.
Borrower can not borrow 40k with 40k collateral.

File: pallets/minterest-protocol/src/lib.rs (hunk 4 above)
Comment (lut1k) [reply pallets/minterest-protocol/src/lib.rs:487]: Fixed

File: pallets/controller/src/mock.rs [hunk 6]
```diff
@@ -85,6 +89,7 @@ parameter_types! {
 impl Config for Runtime {
```
[the remark this answers is not in the document]
Comment (lut1k) [reply pallets/controller/src/mock.rs]: Fixed

File: runtime/src/benchmarking/minterest_protocol.rs (hunk 5 above)
Comment (lut1k) [reply runtime/src/benchmarking/minterest_protocol.rs:35]: Fixed

File: pallets/mnt-token/src/lib.rs [hunk 7]
```diff
@@ -285,67 +282,146 @@ pub mod module {
...
+			total_utility = total_utility.checked_add(utility).ok_or(Error::<T>::NumOverflow)?;
+
+			result.push((currency_id, utility));
+		}
+		Ok((result, total_utility))
+	}
+
+	/// Recalcul`ate MNT speeds
+	pub fn refresh_mnt_speeds() -> DispatchResult {
+		CurrencyId::get_enabled_tokens_in_protocol(UnderlyingAsset)
```
Comment (kurotychb) [inline pallets/mnt-token/src/lib.rs:316]: Are you sure that wee need updating indexes for ALL pools? 
I mean that not all pools can be included to mnt minting.

File: pallets/integration-tests/src/tests/mnt_token_tests.rs [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+			assert_eq!(sum_of_speeds, TestMntToken::mnt_rate());
+		} else {
+			assert_eq!(sum_of_speeds + 1, TestMntToken::mnt_rate());
+		}
+	}
+
+	// This scenario works with two users and three pools.
+	// The test checks the parameters of the MNT token.
+	// Initial parameters: 	DOT + ETH - enabled in mnt minting;
+	// 						mnt_rate = 0.1 MNT per block;
```
[the remark this answers is not in the document]
Comment (lut1k) [reply pallets/integration-tests/src/tests/mnt_token_tests.rs:31]: There are tabs :)

File: pallets/mnt-token/src/lib.rs (hunk 7 above)
Comment (lut1k) [reply pallets/mnt-token/src/lib.rs:316]: FIxed

File: pallets/integration-tests/src/tests/mnt_token_tests.rs [hunk 9]
```diff
@@ -0,0 +1,178 @@
+//  Integration-tests for mnt-token pallet.
```
Comment (DenisRomanovsky) [inline pallets/integration-tests/src/tests/mnt_token_tests.rs:1]:
```suggestion
//  Integration tests for mnt-token pallet.
```

File: pallets/integration-tests/src/tests/mnt_token_tests.rs [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+				assert_ok!(MinterestProtocol::enable_is_collateral(admin(), DOT));
+				assert_ok!(MinterestProtocol::enable_is_collateral(admin(), ETH));
+				assert_ok!(MinterestProtocol::enable_is_collateral(admin(), BTC));
+				assert_ok!(MinterestProtocol::borrow(admin(), DOT, 50_000 * DOLLARS));
+				assert_ok!(MinterestProtocol::borrow(admin(), ETH, 50_000 * DOLLARS));
+				assert_ok!(MinterestProtocol::borrow(admin(), BTC, 50_000 * DOLLARS));
+
+				set_block_number_and_refresh_speeds(10);
+
+				// ALice deposit DOT and enable her DOT pool as collateral.
```
Comment (DenisRomanovsky) [inline pallets/integration-tests/src/tests/mnt_token_tests.rs:76]:
```suggestion
				// Alice deposits DOT and enables her DOT pool as a collateral.
```

File: pallets/integration-tests/src/tests/mnt_token_tests.rs [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+				// (deposit, redeem, borrow, repay, transfer, claim).
+				assert_eq!(TestMntToken::mnt_accrued(ALICE), Balance::zero());
+				assert_eq!(Tokens::free_balance(MNT, &ALICE), Balance::zero());
+
+				set_block_number_and_refresh_speeds(30);
+
+				// There are borrow in all pool, but BTC pool excluded from MNT distribution.
+				test_mnt_speeds(33_333_333_283_333_335, 66_666_666_716_666_664, 0);
+
+				// BOB deposit ETH and enable his assets in pools as collateral.
```
Comment (DenisRomanovsky) [inline pallets/integration-tests/src/tests/mnt_token_tests.rs:95]:
```suggestion
				// BOB deposits ETH and enables his assets in pools as a collateral.
```

File: pallets/integration-tests/src/tests/mnt_token_tests.rs [hunk 12]
```diff
@@ -0,0 +1,178 @@
...
+
+				set_block_number_and_refresh_speeds(60);
+
+				test_mnt_speeds(41_176_429_955_924_386, 58_823_570_044_075_613, 0);
+
+				assert_ok!(MinterestProtocol::borrow(alice(), BTC, 30_000 * DOLLARS));
+
+				set_block_number_and_refresh_speeds(70);
+
+				// The BTC pool is excluded from the MNT-token distribution, so its speed is zero.
```
Comment (DenisRomanovsky) [inline pallets/integration-tests/src/tests/mnt_token_tests.rs:123]:
```suggestion
				// The BTC pool is excluded from the MNT token distribution, so its speed is zero.
```

File: pallets/mnt-token/src/lib.rs [hunk 13]
```diff
@@ -285,67 +282,146 @@ pub mod module {
...
+				.ok_or(Error::<T>::NumOverflow)?;
+
+			total_utility = total_utility.checked_add(utility).ok_or(Error::<T>::NumOverflow)?;
+
+			result.push((currency_id, utility));
+		}
+		Ok((result, total_utility))
+	}
+
+	/// Recalcul`ate MNT speeds
```
Comment (DenisRomanovsky) [inline pallets/mnt-token/src/lib.rs:314]:
```suggestion
	/// Recalculate MNT speeds
```

File: pallets/mnt-token/src/lib.rs [hunk 14]
```diff
@@ -285,67 +282,146 @@ pub mod module {
...
+	/// Transfer MNT tokens to user balance if they are above the threshold.
+	/// Otherwise, put them into internal storage.
+	///
+	/// - `user`: MNT tokens recipient.
+	/// - `user_accrued`: The total amount of accrued tokens.
+	/// - `distribute_all`: boolean, distribute all or part of accrued MNT tokens.
+	fn transfer_mnt(user: &T::AccountId, user_accrued: Balance, distribute_all: bool) -> DispatchResult {
+		//TODO: Need to discuss what we should do.
+		// Erorr/Event/save money to MntAccrued/stop producing mnt tokens
+		let threshold = if distribute_all {
```
Comment (DenisRomanovsky) [inline pallets/mnt-token/src/lib.rs:350]: It should be done through match, shouldn't it?

File: pallets/traits/src/lib.rs [hunk 15]
```diff
@@ -104,9 +104,32 @@ pub trait ControllerAPI<AccountId> {
...
+	// fn get_protocol_interest_threshold(pool_id: CurrencyId) -> Balance;
 
 	// /// Protocol operation mode. In whitelist mode, only members 'WhitelistCouncil' can work with
 	// /// protocols.
 	// fn is_whitelist_mode_enabled() -> bool;
 }
+
+pub trait MntManager<AccountId> {
+	/// Update mnt supply index for pool.
+	/// - `underlying_asset`: The pool whose supply index to update.
```
Comment (DenisRomanovsky) [inline pallets/traits/src/lib.rs:116]:
```suggestion
	/// - `underlying_asset`: The pool which supply index to update.
```

File: pallets/traits/src/lib.rs [hunk 16]
```diff
@@ -104,9 +104,32 @@ pub trait ControllerAPI<AccountId> {
...
-	// fn get_protocol_interest_treshold(pool_id: CurrencyId) -> Balance;
+	// fn get_protocol_interest_threshold(pool_id: CurrencyId) -> Balance;
 
 	// /// Protocol operation mode. In whitelist mode, only members 'WhitelistCouncil' can work with
 	// /// protocols.
 	// fn is_whitelist_mode_enabled() -> bool;
 }
+
+pub trait MntManager<AccountId> {
+	/// Update mnt supply index for pool.
```
Comment (DenisRomanovsky) [inline pallets/traits/src/lib.rs:115]:
```suggestion
	/// Update mnt supply index for a pool.
```

File: pallets/traits/src/lib.rs [hunk 17]
```diff
@@ -104,9 +104,32 @@ pub trait ControllerAPI<AccountId> {
...
+	fn distribute_supplier_mnt(underlying_id: CurrencyId, supplier: &AccountId, distribute_all: bool)
+		-> DispatchResult;
+
+	/// Distribute mnt token to borrower. It should be called after update_mnt_borrow_index.
+	/// Borrowers will not begin to accrue until after the first interaction with the protocol.
+	/// - `underlying_id`: The pool in which the borrower is interacting;
+	/// - `borrower`: The AccountId of the borrower to distribute MNT to.
+	fn distribute_borrower_mnt(underlying_id: CurrencyId, borrower: &AccountId, distribute_all: bool)
+		-> DispatchResult;
+}
```
Comment (DenisRomanovsky) [inline pallets/traits/src/lib.rs:135]:
```suggestion
pub trait MntManager<AccountId> {
	/// Update MNT supply index for a pool.
	/// - `underlying_asset`: The pool which supply index to update.
	fn update_mnt_supply_index(underlying_id: CurrencyId) -> DispatchResult;

	/// Update MNT borrow index for a pool.
	/// - `underlying_asset`: The pool which borrow index to update.
	fn update_mnt_borrow_index(underlying_id: CurrencyId) -> DispatchResult;

	/// Distribute MNT token to supplier. It should be called after update_mnt_supply_index.
	/// - `underlying_id`: The pool in which the supplier is acting;
	/// - `supplier`: The AccountId of the supplier to distribute MNT to.
	fn distribute_supplier_mnt(underlying_id: CurrencyId, supplier: &AccountId, distribute_all: bool)
		-> DispatchResult;

	/// Distribute MNT token to borrower. It should be called after update_mnt_borrow_index.
	/// Borrowers will not begin to accrue tokens till the first interaction with the protocol.
	/// - `underlying_id`: The pool in which the borrower is acting;
	/// - `borrower`: The AccountId of the borrower to distribute MNT to.
	fn distribute_borrower_mnt(underlying_id: CurrencyId, borrower: &AccountId, distribute_all: bool)
		-> DispatchResult;
}
```

File: pallets/integration-tests/src/tests/mnt_token_tests.rs (hunk 9 above)
Comment (lut1k) [reply pallets/integration-tests/src/tests/mnt_token_tests.rs:1]: Fixed

File: pallets/integration-tests/src/tests/mnt_token_tests.rs (hunk 10 above)
Comment (lut1k) [reply pallets/integration-tests/src/tests/mnt_token_tests.rs:76]: Fixed

File: pallets/integration-tests/src/tests/mnt_token_tests.rs (hunk 11 above)
Comment (lut1k) [reply pallets/integration-tests/src/tests/mnt_token_tests.rs:95]: Fixed

File: pallets/integration-tests/src/tests/mnt_token_tests.rs (hunk 12 above)
Comment (lut1k) [reply pallets/integration-tests/src/tests/mnt_token_tests.rs:123]: Fixed

File: pallets/mnt-token/src/lib.rs (hunk 13 above)
Comment (lut1k) [reply pallets/mnt-token/src/lib.rs]: Fixed

File: pallets/mnt-token/src/lib.rs (hunk 14 above)
Comment (lut1k) [reply pallets/mnt-token/src/lib.rs:345]: Fixed

File: pallets/traits/src/lib.rs (hunk 17 above)
Comment (lut1k) [reply pallets/traits/src/lib.rs:135]: Fixed

**Review by kurotychb — APPROVED** [review]

**Review by dev-vadim — APPROVED** [review]

**Review by Apollo50 — APPROVED** [review]

**Pull request merged**